### ISSUE TITLE
Brain Client use of -S

### ISSUE THREAD

**Issue by 0xVavaldi** [hashcat/hashcat#3412]
Hashcat is at https://github.com/hashcat/hashcat/commit/17dcd9ae5207ae40dd017190704098a44d1bf47b so pretty up-to-date

The hashcat brain info posts describes hashcat to be able to run 50kH/s per salt (ofc it depends on hardware) but my limits seemed to be around 1.5-1.7 kH/s per salt on a quick test I did capping at 580MH/s with 313k salts on -m10 (where I can get 15GH/s with -a3) this is feature set 3 so I would like to understand where the difference is and if there are ways to increase this.

The main reason for raising this bug report is brain feature set 2. Feature-set 2 should only report the bare requirements and should not hit any bottlenecks as a result - transferring only the attack info and its progress. However, performance drops significantly with it. Now it is faster by about 300% compared to feature set 3, but at 1.5GH/s it still lags behind the 15GH/s that is possible without --brain-client-features 2.

A quick look into source code might bring about a possible cause. Brain is called every time candidates are distributed, instead of at a set interval causing lag - especially with higher pings - when cycling over it frequently. I believe that the way that the brain client handles and shares candidates and attack info could likely be improved upon - though this is from my limited knowledge & point of view.

I'd love to get some more feedback and thoughts from Atom and other contributors to the brain functionality about this.

**Comment by jsteube** [comment]
Brain functionality automatically sets -S, which has massive impact on performance because all PCI-Express bottleneck workaround by the amplifier is nullified. Try if you achieve similar speed with -S. If yes, please close the ticket because then it's not a bug.

**Comment by 0xVavaldi** [comment]
What effect / issues would happen if we (theoretically) disabled the -S flag while using feature set 2/3?

**Comment by jsteube** [comment]
It's a complex situation, hard to estimate. But I avoid these because when I designed the brain I thought for situation where you want to use the brain, using -S is always a good idea.

**Comment by 0xVavaldi** [comment]
My use case is primarily synchronization of attacks, let's say I want to have a team collaborating with different attacks. They mostly have the same dictionaries and rules available targeting the same list(s). But those lists are NTLM for example, so we use feature set 2 for performance. We don't want to run the same combination of wordlist and rules twice, and if we do we want to be able to finish someone else's attack or help them in its progress (the last one is nice to have. I understand that being very difficult).

But most of all I want
<wordlist> -r <rule list> to never be run twice on the same hashlist. Or if I stop it at 90% that someone can resume from 90 on

**Comment by jsteube** [comment]
I understand, an alternative is using H2P so you can see what your team is doing.

**Comment by 0xVavaldi** [comment]
H2P I presume being Hashtopolis? It has some caveats and downsides - especially with large rule lists and attacks can be run less quickly. No ability to use loopback for example - overall you're missing out on a few functions. Now H2P we do use as well, but I'd love to see if it's possible to get Brain functioning for faster hashes as well.

But what you're saying is that if I remove the -S tag in a custom build of hashcat, it should still work as expected.

**Comment by 0xVavaldi** [comment]
Reference: 
```
- [ Brain Client Features ] -

  # | Features
 ===+========
  1 | Send hashed passwords
  2 | Send attack positions
  3 | Send hashed passwords and attack positions
```
Bumping this to get another look, as there are a few issues with Hashcat Brain that are bothering me. The -S flag is not a hard requirement for the full functionality of Hashcat Brain as I understand and it's causing some issues in the functionality.

As I understand it, -S is needed when candidates must be generated by the CPU before being sent to the Brain server for feature 3. That part makes sense as it will otherwise be computed on GPU. However, feature 2 doesn’t require this since it's the same data held by the status screen. By sending parameters (wordlist, rules, etc.), the server has enough information to determine whether any full candidate or subset of candidates (range) has already been processed. Therefore, -S should not be a requirement for progress updates and shouldn’t be force-enabled for feature set 3.

Another issue is that stdin mode is not supported because -S is forced. I'm not sure what technical limitations require -S other than performance reasons, but perhaps we can find out a way to allow stdin for brain client feature 1 (or 3 as well with the note that it won't send attack positions since that's not possible without additional metadata?)

**Comment by jsteube** [comment]
Yes on point two that is actually possible, but might be technically difficult, especially when it comes to overlapping ranges, like masks. That's why I didn't implement it back then. But with a lot of effort it should be doable. For my taste it's a bit too much effort for what you get from it, but feel free to implement it and send in a PR. I'd be happy to merge
